haystack/nodes/preprocessor/preprocessor.py

```diff
@@ -315,72 +418,103 @@ def split(
         if type(document.content) is not str:
             logger.error("Document content is not of type str. Nothing to split.")
             return [document]
-
+            
+        if pre_split_paragraphs and split_by == "passage":
+            logger.error('"pre_split_paragraphs=True" is not compatible with split_by="passage"')
+            return [document]
         text = document.content
-
+        
+        # Split by paragraph/passage first, if possible. This allows for maximum context when creating embeddings. Mechanism contained in separate private method so that it can be used later if pre_split_paragraphs is set to False
+        if pre_split_paragraphs:
+            paras = self._split_paragraphs(text, merge_short, merge_lowercase)
+        else:
+            paras = [text] # if pre_split_paragraphs is set to False, then we just use the entire document as a single paragraph. Added as a list so it can be used seamlessly by the for loops below
+            
+        # split by words ensuring no sub sentence splits
         if split_respect_sentence_boundary and split_by == "word":
```

Comment:
I think this is a great opportunity to simply this entire if block radically. It's way too complicated for no good reason I think.

I propose replacing lines 434-527 of your version with the following (adjust if needed):

```python
# Feel free to place this check earlier if there isn't any
if split_overlap > split_length:
    raise ValueError("split_lenght must be higher than split_overlap")

splits_pages = []
cur_page = 1

if split_respect_sentence_boundary and split_by == "word":
    
    # added for loop so that we can split by paragraph/passage first,
    # and then by sentence-constrained words within each paragraph/passage. 
    # If pre_split_paragraphs is set to False, then this for loop will only run once
    for para in paras:
        if self.add_page_number:
            # SentenceTokenizer will remove "\f" if it is at the end of a sentence, so substituting it in these
            # cases for "[NEW_PAGE]" to don't lose any page breaks.
            para = self._substitute_page_breaks(para)
        sentences = self._split_sentences(para)

        sentence_buffer = ""
        for sentence in sentences:

            if self.add_page_number and sentence.startswith("[NEW_PAGE]"):
                sentence = sentence.replace("[NEW_PAGE]", "\f")
                cur_page += 1
            
            # Short sentence: put in buffer and continue
            if len(sentence_buffer.split() + sentence.split()) <= split_length:
                sentence_buffer += " " + sentence
                continue

            # Buffer is full: empty it
            if len(sentence_buffer.split()) > split_length:
                # Case of an overfull buffer (can happen with big split_lenght/split_overlap ratios)
                logger.warning(
                    "Found sentence with a word count higher than the split length. "
                    "This can cause latency problems with your Reader. "
                    "The piece (including the overlap buffer) is %s words, %s chars long. First 20 chars: %s", 
                    len(sentence_buffer.split()), len(sentence_buffer), sentence_buffer[:20]
                )
            txt_splits.append(sentence_buffer)
            splits_pages.append(cur_page)
            
            # Compute the overlap and overwrite the buffer
            if split_overlap:
                sentence_buffer = " ".join(sentence_buffer.split()[-split_overlap:])
            else:
                sentence_buffer = ""

            # Add new sentence to buffer
            sentence_buffer += " " + sentence

        if sentence_buffer:
            txt_splits.append(sentence_buffer)
            splits_pages.append(cur_page)
else:
   ...
```

Do you think something important is missing in the above block? I can't see anything else being done there.

I also think this piece of code is well suited for a large array of unit tests. We should add them to ensure that future changes have a good testing baseline to run against. Let me know if you have trouble or questions about adding tests.

Reply:
I've checked the test of the pre-processor and I think you can totally have a look at them, copy a simple one and adapt it to the new algorithm. They're here: https://github.com/deepset-ai/haystack/blob/main/test/nodes/test_preprocessor.py

You might want to have a look at [pytest's documentation](https://docs.pytest.org/en/7.1.x/) if the code looks alien to you. You can also commit one simple test to begin with and I'll give you feedback on it.

Reply:
Ideally it would be every permutation, but of course they can be tons of them. I'd recommend two things:

1. We need to test at least the most common "happy paths" (combinations of parameters that work) and the most common combinations that don't work, to make sure they fail in the way we expect.

2. Look into Pytest parametrization facilities. They might not be what you need, but in some cases they come handy when there are a lot of permutations involved.

3. If you see some corner cases along the way that are not documented or behave in a non-intuitive way, add a test for them too, possibly with a comment explaining why the weird behavior occurs.

You don't need to have a complete set of tests right away. Start with the few most important code paths and expand from there :slightly_smiling_face: